internal/net/call/call.go

```diff
@@ -188,19 +187,29 @@ type serverState struct {
 
 // Serve starts listening for connections and requests on l. Handlers to handle
 // incoming requests are found in hmap.
+//
+// Serve always returns a non-nil error and closes l.
 func Serve(ctx context.Context, l net.Listener, hmap *HandlerMap, opts ServerOptions) error {
 	opts = opts.withDefaults()
 	ss := &serverState{opts: opts}
 	defer ss.stop()
 
-	for ctx.Err() == nil {
+	// Arrange to close the listener when we are canceled.
+	go func() {
+		<-ctx.Done()
+		l.Close()
+	}()
+
+	for {
 		conn, err := l.Accept()
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 		if err != nil {
 			return fmt.Errorf("call server error listening on %s: %w", l.Addr(), err)
 		}
```

Comment:
I removed the condition in the `for` loop and the return at the end of the function and replaced it with this check. I believe this is roughly the same behaviour as before, but is maybe slightly more clear that the function can only return a non-nil error. 

The one difference is that if the context is cancelled while in `serveConnection` there will be one extra call to `l.Accept`, but that should error immediately because `l.Close` would have been called already. I believe it's much more likely that the context is cancelled while in `Accept` because `serveConnection` only blocks for a very short time on a lock to update the set of connections, where as `Accept` could block for much longer waiting on a new connection.

When there are multiple error conditions I generally write them using a switch:

```suggestion
		switch {
		case ctx.Err() != nil:
			return ctx.Err()
		case err != nil:
			return fmt.Errorf("call server error listening on %s: %w", l.Addr(), err)
		}
```

Do you have any preference on the style (one switch vs two if) ?